shared/src/main/scala/async/Async.scala

```diff
@@ -71,13 +78,35 @@ object Async:
           finally lock.unlock()
 
     /** An Async of the same kind as this one, with a new cancellation group */
-    override def withGroup(group: CompletionGroup): Async = Blocking(group)
+    override def withGroup(group: CompletionGroup): Async = Async.LockingAsync(group)
+
+  /** A way to introduce asynchronicity into a synchronous environment. */
+  trait FromSync private[async] ():
+    private[async] type Output[+T]
+    private[async] def apply[T](body: Async ?=> T): Output[T]
+
+  object FromSync:
+    /** A [[FromSync]] implementation that blocks the current runtime. */
+    type Blocking = FromSync { type Output[+T] = T }
+
+    /** Implements [[FromSync]] by directly blocking the current thread. */
+    class BlockingWithLocks(using support: AsyncSupport, scheduler: support.Scheduler) extends FromSync:
+      type Output[T] = T
+      private[async] def apply[T](body: Async.Spawn ?=> T): Output[T] =
+        Async.group(body)(using Async.LockingAsync(CompletionGroup.Unlinked))
 
   /** Execute asynchronous computation `body` on currently running thread. The thread will suspend when the computation
     * waits.
     */
-  def blocking[T](body: Async.Spawn ?=> T)(using support: AsyncSupport, scheduler: support.Scheduler): T =
-    group(body)(using Blocking(CompletionGroup.Unlinked))
+  // def blocking[T](body: Async.Spawn ?=> T)(using blocking: FromSync { type Output[T] = T }): T =
+  //   blocking(body)
+
+  /** Execute asynchronous computation `body` from the context.
+    */
+  def blocking[T](
+      body: Async.Spawn ?=> T
+  )(using fromSync: FromSync): fromSync.Output[T] =
+    fromSync(body)
```

Comment:
Might be worth it to rename `blocking` to `fromSync`, and have `blocking` require a `fromSync: FromSync.Blocking` instance